flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionView.java

```diff
@@ -0,0 +1,199 @@
+			// but also in case the reader is disposed (rather than throwing an exception)
+			return null;
+		}
+
+		updateStatistics(current);
+
+		// We simply assume all the data are non-events for batch jobs to avoid pre-fetching the next header
+		Buffer.DataType nextDataType = numDataAndEventBuffers > 0 ? Buffer.DataType.DATA_BUFFER : Buffer.DataType.NONE;
+		return BufferAndBacklog.fromBufferAndLookahead(current, nextDataType, numDataBuffers, sequenceNumber++);
+	}
+
+	private void updateStatistics(Buffer buffer) {
+		if (buffer.isBuffer()) {
+			numDataBuffers--;
+		}
+		numDataAndEventBuffers--;
+	}
+
+	@Override
+	public boolean isAvailable(int numCreditsAvailable) {
+		// We simply assume there are no events except EndOfPartitionEvent for bath jobs,
+		// then it has no essential effect to ignore the judgement of next event buffer.
+		return numCreditsAvailable > 0 && numDataAndEventBuffers > 0;
+	}
+
+	@Override
+	public void releaseAllResources() throws IOException {
+		// it is not a problem if this method executes multiple times
+		isReleased = true;
+
+		IOUtils.closeQuietly(dataReader);
+
+		// Notify the parent that this one is released. This allows the parent to
+		// eventually release all resources (when all readers are done and the
+		// parent is disposed).
+		parent.releaseReaderReference(this);
+	}
+
+	@Override
+	public boolean isReleased() {
+		return isReleased;
+	}
+
+	@Override
+	public Throwable getFailureCause() {
+		// we can never throw an error after this was created
+		return null;
+	}
+
+	@Override
+	public int unsynchronizedGetNumberOfQueuedBuffers() {
+		return parent.unsynchronizedGetNumberOfQueuedBuffers();
+	}
+
+	@Override
+	public void notifyDataAvailable() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public void resumeConsumption() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public String toString() {
+		return String.format("Blocking Subpartition Reader: ID=%s, index=%d",
+			parent.parent.getPartitionId(),
+			parent.getSubPartitionIndex());
+	}
+
+	/**
+	 * The reader to read from {@link BoundedBlockingSubpartition} and return the wrapped
+	 * {@link org.apache.flink.shaded.netty4.io.netty.channel.FileRegion} based buffer.
+	 */
+	static final class FileRegionReader implements BoundedData.Reader {
+
+		private final FileChannel fileChannel;
+
+		private final ByteBuffer headerBuffer;
+
+		private long lastPosition = -1L;
+
+		private int lastBufferSize;
+
+		FileRegionReader(Path filePath) throws IOException {
+			this.fileChannel = FileChannel.open(filePath, StandardOpenOption.READ);
+			this.headerBuffer = BufferReaderWriterUtil.allocatedHeaderBuffer();
+		}
+
+		@Nullable
+		@Override
+		public Buffer nextBuffer() throws IOException {
+			// As the file region transfer via network channel will not modify the underlining file position,
```

Comment:
Maybe we can optimize this method a bit.

If you inline the code from BufferReaderWriterUtil.readFromByteChannel() here, then you can avoid having the logic tracking `lastPosition` etc, and simply advance the position always here.

It should also help to avoid some calls to `fileChannel.position()`, which are not cheap.

Reply:
The current way is indeed a bit obscure. Actually it is involved in two different usages ATM:

- For network file transfer, the underlining file position for data part is not changed while transferring.

- For local input channel reading, it will change the respective underlining file position via `FileRegionBuffer#readInto`.

So when the `nextBuffer` is called here, it does not know the caller is from local channel or network stack, then it can not always advance the position. Here I rely on the `lastPosition` for implicitly distinguishing the different usages.

My first version PR seems a bit better here to always advance the position in the constructor of extended `FileRegion` instance, since it would be only used for network stack. But after the refactoring in the current version, I did not find another better place to advance the position reasonably. Or I misunderstood your proposed way?

Reply:
I think this is very fragile, using the `lastPosition` to distinguish the uses.

Can we say that this part here (creating regions from the file) always advances the position, and the actual reading part never advances the position? For that, the `FileRegionBuffer#readInto(...)` would also need to take the offset/length fields stored in the region to handle this.

Reply:
If you want, I can try and create a suggestion for this change.

Reply:
Yeah, I agree with the fragile point. Let me consider how to refactor it a bit.

The `FileRegionBuffer#readInto` relies on `FileChannel#read(ByteBuffer dst)` to read the data part, which will automatically updates the position. If I always advance the data part position here before actual reading, then for local channel reading, I need to roll back the respective position. It might be a bit better than current way since we do not need to maintain the `lastPosition` variable. My understanding is right?

The realistic way is to only advance the position for network direct transfer, and leave the normal advance for local input channel while `FileChannel#read(ByteBuffer dst)`. I will further think whether it can be done in some place.